File: Python/arghandler.py
from typing import List

import sys

# =================================================================================================

class CArgValue:
    def __init__ (self, value : str, delims : str = ";,:") -> None:
        self.value = value
        self.subValues = self.Parse (value, delims)

    # The argument parser accepts arguments delimited with ';', ',' or ':'
    # Their hierarchical order is ";.:". 
    # Example: test=a:b:c,d:e:f,g:h:i;1:2:3,4:5:6 would be 
    # "a:b:c,d:e:f,g:h:i" and "1:2:3,4:5:6" on the highest level
    # "a:b:c", "d:e:f", "g:h:i" and "1:2:3", "4:5:6" on the next level
    # arguments are stored recursively
    # delimiters can be skipped (you can e.g. use ';' and ':', but not exchanged)

    def Parse (self, value : str, delims : str) -> list:
        l = len (delims)
        if (l == 0):
            return None
        i = 0
        subValues = []
        while (i < l):
            values = value.split (delims [i])
            if (len (values) > 1):
                for v in values:
                    argV = CArgValue (v, delims [i + 1:])
                    if (argV is not None):
                        subValues.append (argV)
                return subValues
            i += 1
        return None


    def GetVal (self, i : int) -> str:
        return self.value if (self.subValues is None) else self.subValues [i].value

# =================================================================================================

class CArgument ():
    def __init__ (self) -> None:
        self.key = ""
        self.values = None


    def Create (self, arg : str) -> str:
        arg = arg.split ("#")
        values = arg [0].replace ("= ", "=", 1).replace (" =", "=", 1).replace ("\n", "", 1).split ("=")
        self.key = values [0].lower ()
        if (len (values) > 1):
            self.values = CArgValue (values [1])
            # print ("CArgument ({}, {})".format (values [0], values [1]))
        else:
            self.values = CArgValue ("0")
        return self.key
        

    def GetVal (self, i : int = 0) -> str:
        return self.values.GetVal (i)

    # split all arguments into one uniform value array
    # def Parse (self, value, delims):
    #     values = value.split (delims [0])
    #     if (len (delims) == 1):
    #         return values
    #     subValues = []
    #     for v in values:
    #         subValues += self.Parse (v, delims [1:])
    #     return subValues

# =================================================================================================

class CArgHandler ():
    def __init__ (self) -> None:
        self.argList = {}
        for arg in sys.argv:
            self.Add (arg)


    def Add (self, arg):
        a = CArgument ()
        key = a.Create (arg)
        self.argList [key] = a


    def LoadArgs (self, fileName : str = "smileybattle.ini") -> int:
        try:
            with open (fileName, 'r') as f:
                args = f.readlines ()
                f.close ()
        except:
            return -1
        argC : int = 0
        for arg in args:
            if (arg [0] != '#') and (arg [0] != ';'):
                self.Add (arg)
                argC += 1
        return argC


    def GetArg (self, key : str) -> CArgument:
        try:
            a = self.argList [key]
        except KeyError:
            return None
        return a


    def StrVal (self, key, i : int = 0, default : str = None) -> str:
        a = self.GetArg (key)
        if (a is None):
            return default
        return a.GetVal (i)



    def IntVal (self, key, i : int = 0, default : int = None) -> int:
        a = self.GetArg (key)
        if (a is None):
            return default
        return int (a.GetVal (i))


    def FloatVal (self, key, i : int = 0, default : float = None) -> float:
        a = self.GetArg (key)
        if (a is None):
            return default
        return float (a.GetVal (i))


    def BoolVal (self, key, i : int = 0, default : bool = None) -> bool:
        a = self.GetArg (key)
        if (a is None):
            return default
        return bool (int (a.GetVal (i)))

# =================================================================================================

File: CPP/arghandler.cpp
#include <string.h>
#include <stdio.h>
#include <stdlib.h>

#include "textfileloader.h"

#include "argHandler.h"
#include "cavltree.h"

// =================================================================================================

CArgValue::CArgValue(CString value, const char* delims) {
    m_value = value;
    m_subValues = Parse (value, delims);
}


CArgValue::CArgValue(CArgValue const& other) {
    Copy (other);
}


CArgValue& CArgValue::operator= (CArgValue const& other) {
    return Copy (other);
}


CArgValue& CArgValue::Copy (CArgValue const& other) {
    m_value = other.m_value;
    if (other.m_subValues) {
        m_subValues = new CList<CArgValue>;
        *m_subValues = *other.m_subValues;
    }
    else
        m_subValues = nullptr;
    return *this;
}


CArgValue::CArgValue (CArgValue&& other) noexcept {
    Move (other);
}


CArgValue& CArgValue::operator= (CArgValue&& other) noexcept {
    return Move (other);
}


CArgValue& CArgValue::Move (CArgValue& other) {
    m_value = other.m_value;
    m_subValues = other.m_subValues;
    other.m_subValues = nullptr;
    return *this;
}


// The argument parser accepts arguments delimited with ';', ',' or ':'
// Their hierarchical order is ";.:". 
// Example: test=a:b:c,d:e:f,g:h:i;1:2:3,4:5:6 would be 
// "a:b:c,d:e:f,g:h:i" and "1:2:3,4:5:6" on the highest level
// "a:b:c", "d:e:f", "g:h:i" and "1:2:3", "4:5:6" on the next level
// arguments are stored recursively
// delimiters can be skipped (you can e.g. use ';' and ':', but not exchanged)

CList<CArgValue>* CArgValue::Parse(CString value, const char* delims) {
    size_t l = strlen(delims);
    if (l == 0)
        return nullptr;
    CList<CArgValue>* subValues = new CList<CArgValue>;
    for (int i = 0; i < l; i++) {
        CList<CString> values = value.Split(delims[i]);
        if (values.Length() > 1) {
            for (const auto [h, v] : values) {
                CArgValue argV(v, delims + i + 1);
                if (argV.m_value.Length ())
                    subValues->Append(argV);
            }
            return subValues;
        }
    }
    delete subValues;
    return nullptr;
}


CString& CArgValue::GetVal (int i) {
    if (!m_subValues || m_subValues->Empty ())
        return m_value;
    return (*m_subValues)[i].m_value;
}

// =================================================================================================

CString CArgument::Create(CString arg) {
    arg = arg.Split('#')[0];
    CList<CString> values = arg.Replace("= ", "=", 1).Replace(" =", "=", 1).Replace("\n", "", 1).Split('=');
    m_key = values[0].Lower();
    if (values.Length() > 1)
        m_values = CArgValue(values[1]);
    else
        m_values = CArgValue (CString ("0"));
    return m_key;
}
        

CString& CArgument::GetVal(int i) {
    return m_values.GetVal(i);
}

    // split all arguments into one uniform value array
    // void Parse (value, delims) {
    //     values = value.split (delims [0])
    //     if (len (delims) == 1) {
    //         return values
    //     subValues = []
    //     for v in values:
    //         subValues += m_Parse (v, delims [1:])
    //     return subValues

// =================================================================================================

CArgHandler::CArgHandler(int argC, char** argV) {
    m_argList.SetComparator (CString::Compare);
    while (argC--)
        Add(CString (*(argV++)));
}


void CArgHandler::Add(CString arg) {
    CArgument a;
    CString key = a.Create(arg);
    m_argList.Insert(key, a);
}


bool CArgHandler::LineFilter (std::string line) {
    return (line [0] != '#') && (line [0] != ';');
}


int CArgHandler::LoadArgs(const char* fileName) {
    CTextFileLoader f;
    CList<CString>  fileLines;

    int argC = f.ReadLines (fileName, fileLines, LineFilter);
    if (argC > 0)
        for (auto [i, line] : fileLines)
            Add(line);
    return argC;
}


CArgument* CArgHandler::GetArg(const char* key) {
    return m_argList.Find(CString(key));
}


const CString CArgHandler::StrVal(const char* key, int i, CString defVal) {
    CArgument* a = GetArg(key);
    return a ? a->GetVal(i) : defVal;
}


int CArgHandler::IntVal(const char* key, int i, int defVal) {
    CArgument* a = GetArg(key);
    return a ? int (a->GetVal(i)) : defVal;
}


float CArgHandler::FloatVal(const char* key, int i, float defVal) {
    CArgument* a = GetArg(key);
    return a ? float (a->GetVal(i)) : defVal;
}


bool CArgHandler::BoolVal(const char* key, int i, bool defVal) {
    return bool(IntVal(key, i, int(defVal)));
}

CArgHandler* argHandler = nullptr;

// =================================================================================================
